src/main/java/io/supertokens/jwt/JWTSigningKey.java

```diff
@@ -0,0 +1,247 @@
+
+                    // If there are no keys in storage, generate a new one
+                    if (keysFromStorage.isEmpty()) {
+                        try {
+                            keyInfo = generateKeyForAlgorithm(algorithm);
+                            sqlStorage.setJWTSigningKey_Transaction(con, keyInfo);
+                        } catch (NoSuchAlgorithmException | DuplicateKeyIdException | UnsupportedJWTSigningAlgorithmException e) {
+                            throw new StorageTransactionLogicException(e);
+                        }
+                    } else {
+                        // Loop through the keys and find the first one for the algorithm
+                        for (int i = 0; i < keysFromStorage.size(); i++) {
+                            JWTSigningKeyInfo currentKey = keysFromStorage.get(i);
+                            if (currentKey.algorithm.equalsIgnoreCase(algorithm.name())) {
+                                keyInfo = currentKey;
+                                break;
+                            }
+                        }
+
+                        // If no key was found create a new one
+                        if (keyInfo == null) {
+                            try {
+                                keyInfo = generateKeyForAlgorithm(algorithm);
+                                sqlStorage.setJWTSigningKey_Transaction(con, keyInfo);
+                            } catch (NoSuchAlgorithmException | DuplicateKeyIdException | UnsupportedJWTSigningAlgorithmException e) {
+                                throw new StorageTransactionLogicException(e);
+                            }
+                        }
+                    }
+
+                    sqlStorage.commitTransaction(con);
+                    return keyInfo;
+                });
+            } catch (StorageTransactionLogicException e) {
+                if (e.actualException instanceof DuplicateKeyIdException) {
+                    throw (DuplicateKeyIdException) e.actualException;
+                } else if (e.actualException instanceof UnsupportedJWTSigningAlgorithmException) {
+                    throw (UnsupportedJWTSigningAlgorithmException) e.actualException;
+                }
+
+                throw e;
+            }
+        } else if (storage.getType() == STORAGE_TYPE.NOSQL_1) {
+            SessionNoSQLStorage_1 noSQLStorage = (SessionNoSQLStorage_1) storage;
+
+            JWTSigningKeyInfo keyInfo = null;
+
+            while (true) {
+                List<JWTSigningKeyInfo> keysFromStorage = noSQLStorage.getJWTSigningKeys_Transaction();
+
+                // If there are no keys in storage, generate a new one
+                if (keysFromStorage.isEmpty()) {
+                    try {
+                        keyInfo = generateKeyForAlgorithm(algorithm);
+                        boolean success = noSQLStorage.setJWTSigningKeyInfo_Transaction(keyInfo);
```

Comment:
Wouldn't this also throw `DuplicateKeyIdException`?

Reply:
It does, its being handled in the independent function which creates the key now. It loops if that error is thrown and retries